### ISSUE TITLE
test_deepseek41_metal: new router check fails on M1 Ultra (pre-M5); sqrt(softplus) compared at 3e-6 but 1.0f+exp is ~0.7% imprecise

### ISSUE THREAD

**Issue by jasontitus** [antirez/ds4#1039]
**Clarification:** this report is about a failure in the model-free `tests/test_deepseek41_metal` router numerical check on our M1 Ultra. V4.1 inference works on this machine: we have successfully run server requests and inference benchmarks. We are not reporting an inference crash or inability to generate. The test failure also reproduces on pristine upstream `a04f46f`, without our KV compression PR (#186); see the fresh reproduction below in the comments.

---

`tests/test_deepseek41_metal` fails the new router check on M1 Ultra:

```
router n=256 mode=0 rows=1 row=0 expert=4 logit=-11.8886719 actual=0.00263455603 ref=0.00262947031
tests/test_deepseek41_metal.c:105: fabs(probs[t * n + e] - ref[e]) <= 3e-6 * (1 + ref[e])
ds4: Metal cleanup discarded 5 live tensor handles
make: *** [test-deepseek41-metal] Error 1
```

## Repro

```sh
make tests/test_deepseek41_metal && ./tests/test_deepseek41_metal
```

No model or GGUF needed. Fails every run, identical values.

## Environment

| | |
|---|---|
| ds4 commit | `a04f46f` (DeepSeek v4.1 Flash support for CUDA) |
| Hardware | Mac Studio, Apple M1 Ultra, 128 GB |
| OS | macOS 26.6.2 (Darwin 25.6.0) |
| Backend | Metal, `make` default target |
| Metal 4 | unavailable, engine logs `Metal 4 tensor API disabled for pre-M5/pre-A19 devices` |

## Not a regression: the check is new

`check_router` is added wholesale by `a04f46f`; every line of it is new in that commit's diff of `tests/test_deepseek41_metal.c`. Confirmed by bisect on this machine:

| commit | result |
|---|---|
| `bd66c40` DeepSeek v4.1 Flash support for Metal | PASS |
| `a04f46f` DeepSeek v4.1 Flash support for CUDA | FAIL |

So the kernel behaviour is presumably unchanged; the new check is simply tighter than pre-M5 Apple GPUs can satisfy.

## Root cause

The check compares two independent FP32 evaluations of `sqrt(softplus(x))` for a logit of about -11.9, where that expression is ill-conditioned.

Kernel, `metal/dsv4_misc.metal:5188` (and the same form at `:5349`):

```metal
const float4 sp = select(log(1.0f + exp(x)), x, x > 20.0f);
```

Host reference, `tests/test_deepseek41_metal.c:98`:

```c
ref[e] = sqrtf(v > 20 ? (float)v : logf(1.0f + expf((float)v)));
```

For `x = -11.8886719`, `exp(x)` is about 6.87e-6. Adding that to `1.0f` discards most of it: float resolution near 1.0 is 1.19e-7, so the addend carries roughly **0.68% relative error** before the log. Against the true value:

| | value | error vs exact |
|---|---|---|
| exact, `sqrt(log1p(exp(x)))` in double | 0.00262063751 | — |
| host reference, `1.0f + expf` | 0.00262947031 | +8.83e-6 |
| Metal kernel | 0.00263455603 | +1.39e-5 |
| FP32 using `log1pf(expf(x))` | 0.00262063742 | -9.17e-11 |

Both sides are wrong by several times the tolerance, in the same direction, by different amounts. Their gap is 5.09e-6 against a bound of 3.01e-6. The check therefore only passes where the GPU happens to round this expression the same way the host does, which is presumably true on M5 Max and M3 Ultra and is not true here.

Note the second row of that table independently: the router probability the kernel produces is about **0.53% high** for logits near -12. Switching the kernel to `log1p(exp(x))` would remove essentially all of it.

## Suggested fix

Two options, and the choice is yours since they differ in intent:

1. **Fix the reference only.** Use `log1pf(expf(v))` in the test and widen the tolerance enough to cover the kernel's genuine FP32 error at small logits. Keeps kernel behaviour, makes the check portable. The comment at `tests/test_deepseek41_metal.c:96` says the `1.0f +` form was chosen deliberately to match Metal, so this is the smaller change.
2. **Fix the kernel too.** Use `log1p(exp(x))` in `metal/dsv4_misc.metal:5188` and `:5349`, then keep a tight tolerance against an accurate reference. This also removes the 0.53% routing-probability error above.

## Scope of verification

Verified on Metal, M1 Ultra only. CUDA and ROCm untested. I do not have an M5 or M3 to confirm the check passes there.

**Comment by lin16303** [comment]
m1ultra works fine.
 ./ds4-server --ssd-streaming --ctx 32768  -m DeepSeek-V4.1-Flash-Q2.gguf
0914 19:15:21 ds4-server: live kv cache miss live=22253 prompt=646 common=2 vision=match reason=token-mismatch
0914 19:15:21 ds4-server: chat ctx=0..646:646 prompt start
0914 19:15:54 ds4-server: chat ctx=0..646:646 prefill chunk 646/646 (100.0%) chunk=0.00 t/s avg=19.24 t/s 33.577s
0914 19:15:55 ds4-server: chat ctx=0..646:646 prompt done 33.665s
0914 19:16:00 ds4-server: chat ctx=646..696:50 gen=50 decoding chunk=9.72 t/s avg=9.72 t/s 5.146s
0914 19:16:04 ds4-server: chat ctx=696..737:41 gen=91 decoding chunk=10.35 t/s avg=9.99 t/s 9.108s
0914 19:16:04 ds4-server: chat ctx=0..646:646 gen=91 finish=stop 42.773s
0914 19:16:21 ds4-server: live kv cache miss live=737 prompt=3203 common=81 vision=match reason=token-mismatch
0914 19:16:21 ds4-server: chat ctx=0..3203:3203 prompt start
0914 19:16:55 ds4-server: chat ctx=0..3203:3203 prefill chunk 2048/3203 (63.9%) chunk=0.00 t/s avg=59.18 t/s 34.605s
0914 19:17:27 ds4-server: chat ctx=0..3203:3203 prefill chunk 3203/3203 (100.0%) chunk=36.20 t/s avg=48.16 t/s 66.508s
0914 19:17:27 ds4-server: chat ctx=0..3203:3203 prompt done 66.597s
0914 19:17:32 ds4-server: chat ctx=3203..3253:50 gen=50 decoding chunk=9.58 t/s avg=9.58 t/s 5.217s
0914 19:17:37 ds4-server: chat ctx=3253..3300:47 gen=97 decoding chunk=10.28 t/s avg=9.91 t/s 9.788s
0914 19:17:37 ds4-server: chat ctx=0..3203:3203 gen=97 finish=stop 76.385s
^C0914 19:19:59 ds4-server: shutdown requested, draining requests

**Comment by emilianbold** [comment]
> ./tests/test_deepseek41_metal ... presumably true on M5 Max

I can confirm the test passes on M5 Max.

I'm also curious if the attempted fix in https://github.com/antirez/ds4/pull/1044 works for you.

**Comment by jasontitus** [comment]
Thanks for the reports. To clarify: **V4.1 inference works on our M1 Ultra too. This issue is about the failing router numerical assertion in `tests/test_deepseek41_metal`, not a failure to run inference.** I have added that distinction at the top of the issue.

I also rechecked that our KV compression PR (#186) is unrelated to the reproduction:

- Exported pristine upstream `a04f46fa423e45712c8c7e430eff422479f314a3` into a fresh directory with `git archive`.
- Built `tests/test_deepseek41_metal` from scratch, with no reused object files or compression sources.
- Removed all `DS4_*` environment overrides and ran the model-free test three times on M1 Ultra / macOS 26.6.2.

All three runs failed identically at line 105:

```text
router n=256 mode=0 rows=1 row=0 expert=4 logit=-11.8886719 actual=0.00263455603 ref=0.00262947031
```

Successful server inference is consistent with this narrower numerical-test failure. Thanks also for confirming the test passes on M5 Max and for linking #1044; I have not tested that proposed fix on this M1 Ultra yet.

**Comment by jasontitus** [comment]
@emilianbold Yes, #1044 fixes it on M1 Ultra. The router test fails 3/3 on its base `9139e2a` and passes 3/3 with the PR, including with the pre-M5 fusion disabled and in math-safe mode. Full results are on the PR.

**Comment by emilianbold** [comment]
@jasontitus the fix has been merged into main branch. If it works for you please close the issue.

**Comment by jasontitus** [comment]
Confirmed fixed on current main (0aaea5a), which includes 0719a0b from #1044. Rebuilt with make tests/test_deepseek41_metal and ran the full model-free binary three times on our Apple M1 Ultra (128 GB), with all DS4_* environment overrides removed. All three runs exited 0; the previously failing check now reports: 256/384-expert routing, ties/extremes and scalar/batch oracle: PASS. Thanks @emilianbold for the fix, and thanks for getting it merged! Closing as resolved.
